clippy_lints/src/future_not_send.rs

```diff
@@ -0,0 +1,89 @@
+use crate::utils;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{Body, FnDecl, HirId};
+use rustc_infer::infer::TyCtxtInferExt;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::Opaque;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::{Span, sym};
+use rustc_trait_selection::traits::{self, FulfillmentError, TraitEngine};
+use rustc_trait_selection::traits::error_reporting::InferCtxtExt;
+
+declare_clippy_lint! {
+    /// **What it does:** This lint requires Future implementations returned from
+    /// functions and methods to implement the `Send` marker trait.
+    ///
+    /// **Why is this bad?** Returning a non-Send Future makes this Future
+    /// useless to code that uses multithreaded executors; even worse, the
+    /// error will typically not be reported in the source file where the
+    /// problem lies. Do not use this lint for single-threade (e.g. embedded)
+    /// targets.
```

Comment:
Sorry, if that is a stupid question (I'm not at all familiar with async/await programming). You're talking here about returning a non-send Future, but looking at the examples shows functions, that doesn't return anything (well `Future<()>` I guess). Also the tests refer to the function arguments most of the time and say that they have to be `Send` or `Sync`. 

So... I don't even really know, what my question here is. But as an `async`/`await` noob, I don't know, what this lint tries to tell me and why this is bad.

Reply:
It's not about the `()` *final produced value* but all of the *state* kept around *until that value is produced*. The future may be executed in a thread-pool which requires the ability to move that state across threads at will.

Reply:
Or answered differently: I’m not sure how much of an intro to async/await should be included with the documentation of this lint. It should probably say something about “enable only if you are using async/await” or similar.

Reply:
@eddyb explanation made this clear to me, thanks! Something like this should be included in the `Why this is bad` explanation. An intro to async/await is not necessary.